### ISSUE TITLE
Trailing comma in JSON breaks the whole thing

### ISSUE THREAD

**Issue by tomassasovsky** [rknell/alfred#85]
the body I'm sending:
```JSON
{
    "email": "to.sasovsky@gmail.com",
}
```
the error I'm getting:

![image](https://user-images.githubusercontent.com/63405252/160504979-0aba143d-e9f2-48c6-9491-1c7c3642e123.png)

**Comment by rknell** [comment]
Ok, so at the end of the day, its invalid JSON. Darts JSON decoder falls over:
<img width="912" alt="Screen Shot 2022-03-29 at 10 20 29 am" src="https://user-images.githubusercontent.com/5875053/160508318-f461aa37-f9a5-42d8-9b91-9e4d5701e5d6.png">

and even the javascript JSON decoder will throw if you put a trailing comma:

<img width="428" alt="Screen Shot 2022-03-29 at 10 17 50 am" src="https://user-images.githubusercontent.com/5875053/160508107-9690c19c-7951-4ca3-8aea-8f7ed6f113ca.png">

So I don't think it should handle it gracefully. However - it should throw an error that its an invalid body, as opposed to header already sent. So I will do some work on throwing a better error but will leave the invalid JSON alone.

**Comment by tomassasovsky** [comment]
I understand, the thing is that it takes the app down. I've tried adding the `onInternalError` function but still.

**Comment by rknell** [comment]
we definitely don't want to crash the server!

**Comment by rknell** [comment]
I just went to have a look at it and the most recent test is there to handle invalid JSON input and return a 400 (thereby not crashing the server) what version of Alfred are you using?

**Comment by tomassasovsky** [comment]
using `alfred: ^0.1.5+3`

**Comment by rknell** [comment]
That release was literally the one that added the test for dodgy JSON so it should be working.

It might be that the published version doesn't include the code - ie I made a mistake. Do you mind adding a dependency for the master branch and seeing if that resolves it? If it does I will release a new version.

```pubspec.yaml
alfred:
    git: https://github.com/rknell/alfred.git
```

**Comment by tomassasovsky** [comment]
tried that, I'm getting a pretty nasty error:
![image](https://user-images.githubusercontent.com/63405252/160513015-b0d21594-6736-4921-bba1-2ba662fbf9bd.png)

**Comment by tomassasovsky** [comment]
Tried running it directly with dart, but still..
![image](https://user-images.githubusercontent.com/63405252/160513108-b8adf39c-d2cb-4728-856b-32f92e979efa.png)

**Comment by rknell** [comment]
Ahhhh....

So I think its being caused by you using a try catch block and then trying to return something.

I can reproduce it with this:

```dart
  test('it handles invalid json input with a trailing comma', () async {
    app.post('/test', (req, res) async {
      try {
        await req.body;
      } catch (e) {
        throw AlfredException(500, {'test': 'response'});
      }
    });
    final response = await http.post(Uri.parse('http://localhost:$port/test'),
        body: '{ "email": "to.sasovsky@gmail.com",}',
        headers: {'Content-Type': 'application/json'});
    expect(response.statusCode, 400);
  });
```

I'll see if there is a way not to crash the server when you have already sent a header.

**Comment by seceba** [comment]
Thank you for your help, you have shared very useful information.

I would like to make one addition. If we add double quotes to the key name when submitting Form-Data and we use AlfredException to catch the error, alfred crashes.


Postman:

curl --location --request DELETE '127.0.0.1:3000/testnew' \
--form '"fooo"="123"'

**Key name "foo" with quotes.**

Endpoint Codes:

server.post("/testnew", (req, res) async {
      try {
        var testbody = await req.body;
        return testbody.toString();
      } catch (e) {
        throw AlfredException(401, "Error");
      }
    });

Crash Errors:

**Unhandled exception:
Bad state: Header already sent
#0      _HttpResponse.statusCode= (dart:_http/http_impl.dart:1190:35)
#1      Alfred._incomingRequest
package:alfred/src/alfred.dart:364
<asynchronous suspension>
#2      _QueuedFuture.execute
package:queue/src/dart_queue_base.dart:26
<asynchronous suspension>**

**Comment by rknell** [comment]
should be resolved in v0.1.6+1
